scripts/nb-tester/qiskit_docs_notebook_tester/post_process.py

```diff
@@ -37,15 +43,43 @@ def post_process_notebook(nb: nbformat.NotebookNode) -> nbformat.NotebookNode:
         # Remove execution metadata to avoid noisy diffs.
         cell.metadata.pop("execution", None)
         if "version-info" in cell.metadata.get("tags", []):
-            cell.source = VERSION_INFO.format(packages=get_package_versions())
+            python_code = "\n".join(cell.source for cell in nb.cells if cell.cell_type == 'code')
+            requirements_txt = Path("scripts/nb-tester/requirements.txt").read_text()
+            cell.source = VERSION_INFO.format(packages=get_package_versions(python_code, requirements_txt))
 
     nb, _ = clean_notebook(nb)
     return nb
 
 
-def get_package_versions():
-    requirements_file = Path("scripts/nb-tester/requirements.txt").read_text()
+def get_package_versions(python_code: str, requirements_txt: str) -> str:
+    # A module can refer to many installable packages, e.g. 'import qiskit' can import
+    # things installed by both 'qiskit' and 'qiskit[visualization]'. For simplicity,
+    # we include any packages that could be relevant.
+    module_to_packages = importlib.metadata.packages_distributions()
+    packages = flatten_to_list([
+      module_to_packages[module]
+      for module in get_used_modules(python_code)
+      if module not in sys.stdlib_module_names
+    ])
     package_versions = "\n".join(
-        line for line in requirements_file.split("\n") if not line.startswith("#")
+        line for line in requirements_txt.split("\n")
+        if re.split('[\\[~=]', line)[0].strip() in packages
     )
     return package_versions.strip()
+
+def get_used_modules(source: str) -> Iterable[str]:
+    # Remove Jupyter magics
+    source = "\n".join([
+        line for line in source.split("\n")
+        if not line.strip().startswith("%")
+    ])
+    for node in ast.iter_child_nodes(ast.parse(source=source, filename="_.py")):
+        if isinstance(node, ast.Import):
+            for module in node.names:
+                yield module.name.split(".")[0]
+        elif isinstance(node, ast.ImportFrom):
+            if node.module is not None:
+                yield node.module.split(".")[0]
+
+def flatten_to_list(i: Iterable[Iterable]) -> list:
```

Comment:
Imo it's fine to call this `flatten` because the type hint already makes clear it's a list and it's a sensible default. I wish Python std lib implemented it.